src/encrypted_container.rs

```diff
@@ -0,0 +1,248 @@
+    fn try_from(value: EncryptedDataContainer) -> Result<Self> {
+        let mut buf = [0u8; SERIALIZED_OBJECT_BUFFER_SIZE];
+        let r = cbor_serialize(&value, &mut buf).map_err(|_| Error::ObjectSerializationError)?;
+        let bytes = Message::from_slice(r).map_err(|_| Error::SerializationBufferTooSmall)?;
+        Ok(bytes)
+    }
+}
+
+impl EncryptedDataContainer {
+    /// Decrypt given Bytes and return original object instance
+    pub fn decrypt_from_bytes<T, O>(
+        trussed: &mut T,
+        ser_encrypted: Message,
+        encryption_key: KeyId,
+    ) -> Result<O>
+    where
+        T: trussed::Client + trussed::client::Chacha8Poly1305,
+        for<'a> O: Deserialize<'a>,
+    {
+        let deserialized_container: EncryptedDataContainer =
+            cbor_deserialize(&ser_encrypted).map_err(|_| Error::DeserializationToContainerError)?;
+
+        deserialized_container.decrypt(trussed, None, encryption_key)
+    }
+
+    /// Create Encrypted Data Container from the given object
+    pub fn from_obj<T, O>(
+        trussed: &mut T,
+        obj: &O,
+        associated_data: Option<&[u8]>,
+        encryption_key: KeyId,
+    ) -> Result<EncryptedDataContainer>
+    where
+        T: trussed::Client + trussed::client::Chacha8Poly1305,
+        O: Serialize,
+    {
+        let mut buf = [0u8; SERIALIZED_OBJECT_BUFFER_SIZE];
+        let message = cbor_serialize(obj, &mut buf).map_err(|_| Error::ObjectSerializationError)?;
+        debug_now!("Plaintext size: {}", message.len());
+        Self::encrypt_message(trussed, message, associated_data, encryption_key)
+    }
+
+    /// Encrypt given Bytes object, and return an Encrypted Data Container
+    pub fn encrypt_message<T>(
+        trussed: &mut T,
+        message: &[u8],
+        associated_data: Option<&[u8]>,
+        encryption_key: KeyId,
+    ) -> Result<EncryptedDataContainer>
+    where
+        T: trussed::Client + trussed::client::Chacha8Poly1305,
+    {
+        #[cfg(feature = "no-encrypted-credentials")]
+        {
+            // Skipping error handling, as this feature is only for the debugging purposes
+            return Ok(EncryptedDataContainer {
+                data: Message::from_slice(&message).unwrap(),
+                nonce: Default::default(),
+                tag: Default::default(),
+                // nonce: ShortData::from_slice(&[12u8, 12]).unwrap(),
+                // tag: ShortData::from_slice(&[16u8, 16]).unwrap(),
+            });
+        }
+
+        // nonce is provided internally via internal per-key counter, hence not passed here
+        let encryption_results = try_syscall!(trussed.encrypt_chacha8poly1305(
+            encryption_key,
+            message,
+            associated_data.unwrap_or_default(),
+            None
+        ))
+        .map_err(|_| Error::FailedEncryption)?;
+
+        let ciphertext = Message::from_slice(&encryption_results.ciphertext)
+            .map_err(|_| Error::FailedContainerSerialization)?;
```

Comment:
Can’t we just move `ciphertext` out of `encryption_results` instead of copying?

Reply:
Good point